tests/net/mod.rs

```diff
@@ -990,24 +986,42 @@ where
 
 impl<C, D, N> VirtualNet<D>
 where
-    D: DistAlgorithm<Output = Batch<C, N>>,
+    D: DistAlgorithm<NodeId = N, Output = Batch<C, N>>,
     C: Contribution + Clone,
     N: NodeIdT,
 {
-    /// Verifies that all nodes' outputs agree, and returns the output.
-    pub fn verify_batches(&self) -> &[Batch<C, N>] {
-        let first = self.correct_nodes().nth(0).unwrap().outputs();
-        let pub_eq = |(b0, b1): (&Batch<C, _>, &Batch<C, _>)| b0.public_eq(b1);
-        for (i, node) in self.correct_nodes().enumerate().skip(0) {
+    /// Verifies that all nodes' outputs agree, given a correct "full" node that output all
+    /// batches in a total order and with no gaps.
+    ///
+    /// The output of the full node is used to derive in expected output of other nodes in every
+    /// epoch. After that the check ensures that correct nodes output the same batches in epochs
+    /// when those nodes were participants (either validators or candidates).
+    pub fn verify_batches<E>(&self, full_node: &Node<D>)
+    where
+        Batch<C, N>: SenderQueueableOutput<N, E>,
+    {
+        let mut participants: BTreeSet<N> = self.nodes().map(Node::id).cloned().collect();
+        let mut expected: BTreeMap<N, Vec<_>> = BTreeMap::new();
+        for batch in &full_node.outputs {
+            for id in &participants {
+                expected.entry(id.clone()).or_default().push(batch);
+            }
+            if let Some(new_participants) = batch.participant_change() {
+                participants = new_participants;
+            }
```

Comment:
This wouldn't work with a pure observer node, I think?
If we don't need that right now, I'd add a `TODO` here: Ideally we'd need the same logic as in the `SenderQueue` implementation itself, where we remove the _difference_ of the previous batch's and the new participants, etc.

Reply:
Please check if this comment still applies. I'm not sure about the "pure observer". The same procedure works with a usual observer in `TestNetwork::verify_batches`.

Reply:
But does `TestNetwork::verify_batches` still work if you add comparing the _length_ of the observer's output to the expected one?

Here I'm just thinking of the case where one of `self.nodes()` is an observer from the beginning. It would get removed from `participants` here with the first batch that contains a change.
I guess we don't have that in the tests at the moment, but I'd maybe add a `TODO` here to account for observers.

Reply:
That observer would never be a participant. So it won't be removed here. The way `TestNetwork` handles observers is by explicitly adding those to the peer list of the sender queue.